scapy/interfaces.py

```diff
@@ -0,0 +1,347 @@
+# Copyright (C) Gabriel Potter <gabriel@potter.fr>
+# This program is published under a GPLv2 license
+
+"""
+Interfaces management
+"""
+
+import itertools
+import uuid
+from collections import defaultdict
+
+from scapy.config import conf
+from scapy.consts import WINDOWS
+from scapy.utils import pretty_list
+
+from scapy.modules.six.moves import UserDict
+import scapy.modules.six as six
+
+
+class InterfaceProvider(object):
+    name = "Unknown"
+    headers = ("INDEX", "IFACE", "IPv4", "IPv6", "MAC")
+    header_sort = 3
+    libpcap = False
+
+    def load(self):
+        """Returns a dictionary of the loaded interfaces, by their
+        name."""
+        raise Exception("load() should be implemented !")
+
+    def reload(self):
+        """Same than load() but for reloads. By default calls load"""
+        return self.load()
+
+    def l2socket(self):
+        """Return L2 socket used by interfaces of this provider"""
+        return conf.L2socket
+
+    def l2listen(self):
+        """Return L2listen socket used by interfaces of this provider"""
+        return conf.L2listen
+
+    def l3socket(self):
+        """Return L3 socket used by interfaces of this provider"""
+        return conf.L3socket
+
+    def _is_valid(self, dev):
+        """Returns whether an interface is valid or not"""
+        return dev.valid
+
+    def _format(self, dev, **kwargs):
+        """Returns a tuple of the elements used by show()"""
+        mac = dev.mac
+        resolve_mac = kwargs.get("resolve_mac", True)
+        if resolve_mac and conf.manufdb:
+            mac = conf.manufdb._resolve_MAC(mac)
+        index = str(dev.index)
+        return (index, dev.description, str(dev.ip), str(dev.ip6), mac)
+
+
+class NetworkInterface(object):
+    def __init__(self, provider, data=None):
+        self.provider = provider
+        self.name = ""
+        self.description = ""
+        self.network_name = ""
+        self.index = -1
+        self.ip = None
+        self.ip6 = None
+        self.ips = []
+        self.mac = None
+        self.valid = True
+        self.dummy = False
+        if data is not None:
+            self.update(data)
+
+    def update(self, data):
+        """Update info about a network interface according
+        to a given dictionary. Such data is provided by providers
+        """
+        self.name = data.get('name', "")
+        self.description = data.get('description', "")
+        self.network_name = data.get('network_name', "")
+        self.index = data.get('index', 0)
+        self.ip = data.get('ip', "")
+        self.ip6 = data.get('ip6', "")
+        self.ips = data.get('ips', [])
+        self.mac = data.get('mac', "")
+        self.flags = data.get('flags', 0)
+        self.valid = data.get('valid', True)
+        self.dummy = data.get('dummy', False)
+
+        # Backup to chose main IPs
+        if self.ips:
+            try:
+                self.ip = next(x for x in self.ips if ":" not in x)
```

Comment:
For IPv6, this code probably needs improvement (there will probably be multiple IPv6 addresses. What should be considered the defaul tone?)

Reply:
Difficult question. It will depend on the routing table, and we will need to fallback to use the same information as the 'Src candidates' output of `conf.routes6`. I think that it will be better to try to display all of them, or have a crop mechanism such as done with `conf.routes6`.